File: createRepo.go
/* 
Rob Casale
This will:
 - Validate user does exist and if it does not throw Error
 - Validate repo does not exist and if it does throw Error
 - Create Repo
Return codes:
 - 0 success
 - 1 user does not exist
 - 2 repo exist
 - 3 error create repo
 - 8 others
*/

package main

import (
	"fmt"				// used for printing messages
	"os"				// used to setup proxy settings
	"io/ioutil"			// used to read in files and http responses
    "net/http"			// used for http requests
    "crypto/tls"		// used for certificate
    "crypto/x509"		// used for certificate
    "bytes"				// used to send json messages
)

func main() {
	// admin DTR credentials
	admin_username := "casaler"
	admin_password := "adpadp11"
	
	// DTR fully qualified hostname
	dtr_full_hostname := "cdldvostk1vdtr001.es.ad.adp.com"
	
	// API responses
	success := "201 Created"
	account_exists := "200 OK"
	repo_exists := "400 Bad Request"
	//account_does_not_exist := "404 Not Found"
	//json_not_specified := "415 Unsupported Media Type"
	
	// exit statuses
	exit_success := "exit(0): success"
	exit_user_does_not_exist := "exit(1): username does not exists"
	exit_repo_exists := "exits(2): repository already exists"
	//exit_error_create_user := "exit(4): could not create username"
	//exit_error_create_repo := "exit(5): could not create repository"
	exit_other := "exit(8): unknown error"
	incorrect_arguments := "exit(8): provide username and password"
	
	
	// Set proxy settings
	setupProxy()
	
	// retrieve command line arguments
	if len(os.Args) == 3 {
		username := os.Args[1]
		image := os.Args[2]
		
		// Request urls
		account_details_url := "https://"+dtr_full_hostname+"/api/v0/accounts/"
		create_repo_url := "https://"+dtr_full_hostname+"/api/v0/repositories/"+username
		
		// official version - var jsonCreateRepo = []byte(`{"name": "string","shortDescription": "string","longDescription": "string","visibility": "public"}`)
		var jsonCreateRepo = []byte(`{"name": "`+image+`","visibility":"public"}`)
		
		response, err := getContent(account_details_url+username,admin_username,admin_password)
		if err != nil {
	    	fmt.Println("Did not get a response from the url")
		} else {
	    	// the given username exists
	    	if response.Status == account_exists {
	    		// create repo
	    		repo_response, err := postContent(create_repo_url,admin_username,admin_password,jsonCreateRepo)
	    		if err != nil {
	    			fmt.Println("Did not get a response from the url")
	    		} else {
	    			if repo_response.Status == success {
	    				fmt.Println(exit_success)
	    			} else if repo_response.Status == repo_exists {
	    				// could not create repo
	    				fmt.Println(exit_repo_exists)
	    			} else {
	    				// other errors
	    				fmt.Println(exit_other)
	    			}
	    		}
	    	} else {
	    		// username does not exist, cannot create repo
	    		fmt.Println(exit_user_does_not_exist)
	    	}
		}
	} else {
		fmt.Println(incorrect_arguments)
	}
}

/**
 * This will set all of the proxy variables that are needed to get through
 * to the other side
 * @param None
 * @return None
 */
func setupProxy() {
	// By-pass proxy
	os.Setenv("HTTP_PROXY", "http://paascloud:8e)P4m)I5y(F9@usproxy.es.oneadp.com:8080")
	os.Setenv("HTTPS_PROXY", "http://paascloud:8e)P4m)I5y(F9@usproxy.es.oneadp.com:8080")
	os.Setenv("NO_PROXY", ".es.oneadp.com")
	os.Setenv("http_proxy", "http://paascloud:8e)P4m)I5y(F9@usproxy.es.oneadp.com:8080")
	os.Setenv("https_proxy", "http://paascloud:8e)P4m)I5y(F9@usproxy.es.oneadp.com:8080")
	os.Setenv("no_proxy", ".es.oneadp.com")
}

/**
 * This will create a client will the DTRs self-signed certificate such that
 * API calls can be made
 * @param None
 * @return client with the certificate added
 */
func createClient() (*http.Client) {
	cert := x509.NewCertPool()
    dtr_cert, err := ioutil.ReadFile("ca.crt")
    if err != nil {
    	fmt.Println("Could not read in certificate")
    	return nil
    }
	cert.AppendCertsFromPEM(dtr_cert)
	tr := &http.Transport{
		TLSClientConfig:    &tls.Config{RootCAs: cert},
		DisableCompression: true,
	}
	client := &http.Client{Transport: tr}
	return client
}

/**
 * This function will perform a GET request on the url you specified 
 * @param request url
 * @return array of bytes if retrieved successfully, otherwise the function
 * will return an error
 */
func getContent(url string, admin_username string, admin_password string) (*http.Response, error) {
    // Build the request
    req, err := http.NewRequest("GET", url, nil)
    if err != nil {
      fmt.Println("Build failed: ",err)
      return nil, err
    }
    req.SetBasicAuth(admin_username,admin_password)
    // Send the request via a client
    resp, err := createClient().Do(req)
    if err != nil {
      fmt.Println("Send request failed: ",err)
      return nil, err
    }
    // Defer the closing of the body
    defer resp.Body.Close()
    return resp, nil
}

/**
 * This function will perform a POST request on the url you specified 
 * @param request url
 * @return array of bytes if retrieved successfully, otherwise the function
 * will return an error
 */
func postContent(url string, admin_username string, admin_password string,
	jsonMessage []byte) (*http.Response, error) {
	// Build the request
	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonMessage))
	if err != nil {
      fmt.Println("Build failed: ",err)
      return nil, err
    }
	req.SetBasicAuth(admin_username,admin_password)
	req.Header.Set("Content-Type", "application/json")
	// Send the request via a client
	resp, err := createClient().Do(req)
	if err != nil {
      fmt.Println("Send request failed: ",err)
      return nil, err
    }
	// Defer the closing of the body
	defer resp.Body.Close()
	return resp, nil
}

File: createRepo.py
# Rob Casale
# This will:
# - Validate user does exist and if it does not throw Error
# - Validate repo does not exist and if it does throw Error
# - Create Repo
# Return codes:
# - 0 success
# - 1 user does not exist
# - 2 repo exist
# - 3 error create repo
# - 8 others

import requests, os, ssl, urllib2, base64, json, sys

# Another possible context option
# context=ssl._create_unverified_context()

# Get pass verification and get response
gcontext = ssl.SSLContext(ssl.PROTOCOL_TLSv1)

# API url requests
test_url='https://www.google.com'
url_account='https://cdldvostk1vdtr001.es.ad.adp.com/api/v0/accounts/'
url_repo='https://cdldvostk1vdtr001.es.ad.adp.com/api/v0/repositories/'

# Login credentials
username='casaler'
password='adpadp11'

# Proxy settings
proxies={
  "http":"http://paascloud:8e)P4m)I5y(F9@usproxy.es.oneadp.com:8080",
  "https":"http://paascloud:8e)P4m)I5y(F9@usproxy.es.oneadp.com:8080",
}

def build_request(requestUrl):
	# Build request
	request=urllib2.Request(requestUrl)
	base64string = base64.encodestring('%s:%s' % (username, password)).replace('\n', '')
	request.add_header("Authorization", "Basic %s" % base64string)
	return request

def get_info(request):
	info=urllib2.urlopen(request,context=gcontext).read()
	return info

def get_code(request):
        code=urllib2.urlopen(request,context=gcontext).getcode()
        return code

def create_account_code(request,user):
        data={'type':'user','name':user,'password':password}
	request.add_header('Content-Type', 'application/json')
	code=urllib2.urlopen(request,json.dumps(data),context=gcontext).getcode()
        return code

def activate_account_code(request,user):
	request.add_header('Content-Type', 'application/json')
	request.get_method=lambda:'PUT'
        code=urllib2.urlopen(request,json.dumps(user),context=gcontext).getcode()
        return code

def create_repository_code(request,repo,visibility):
	data={'name':repo,'visibility':visibility}
        request.add_header('Content-Type', 'application/json')
        code=urllib2.urlopen(request,json.dumps(data),context=gcontext).getcode()
        return code

# Retrieve username and repo from user
if len(sys.argv) != 3:
	print 'exit(8): enter username and repository only'
	exit()

user=sys.argv[1]
repo=sys.argv[2]

user_code=None
url_user_account=url_account+user

try:
	# validate username
	user_code=get_code(build_request(url_user_account))
	repo_code=None
	url_user_repo=url_repo+user+'/'+repo

	try:
		# validate repository
		repo_code=get_code(build_request(url_user_repo))
		print 'exit(2): repository already exist'

	except urllib2.HTTPError as e:
		if e.code == 404:
			# repo does not exist
			#visibility=raw_input('Enter the visibility of the repository[public/private]: ').lower()
			visibility='public'
			try:
				# create repo
				create_repository_code(build_request(url_repo+user),repo,visibility)
				print 'exit(0): success'
			
			except urllib2.HTTPError as e:
				if e.code == 400:
					# repo exists - should not happen
					print 'exit(2): repository already exist'
				else:
					# could not create repositroy
					print 'exit(3): could not create repository'
		else:
			# could not access repo details
			print 'exit(8): could not access repository details'

except urllib2.HTTPError as e:
	if e.code == 404:
		# username does not exist
		print 'exit(1): username does not exist'
	else:
		# could not access account details
		print 'exit(8): could not access account details'


